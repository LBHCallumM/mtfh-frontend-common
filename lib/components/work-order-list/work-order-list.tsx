import React from "react";
import { useWorkOrders } from "../../api/work-order/v2";
import { config } from "../../config";
import locale from "../../locale";
import { CardList } from "../card-list";
import { Center } from "../center";
import { ErrorSummary } from "../error-summary";
import { Link } from "../link";
import { Spinner } from "../spinner";
import WorkOrderListItem from "./work-order-list-item";

interface WorkOrderListProps {
  assetId: string;
}

export const WorkOrderList = ({ assetId }: WorkOrderListProps) => {
  const { data: workOrders, error } = useWorkOrders(assetId);
  const { components } = locale;

  if (error) {
    return (
      <ErrorSummary
        id="work-order-list-error"
        title={components.workOrderList.errors.unableToFetchWorkOrder}
        description={components.workOrderList.errors.unableToFetchWorkOrderDescription}
      />
    );
  }

  if (!workOrders) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  return (
    <div>
      <CardList>
        {workOrders.map((workOrder, index) => (
          <WorkOrderListItem key={index} workOrder={workOrder} />
        ))}
      </CardList>
      <Link
        href={`${config.repairsHubAppUrl}/properties/${assetId}`}
        isExternal
        className="repair-list__link"
      >
        {components.workOrderList.seeAllWorkOrders}
      </Link>
    </div>
  );
};
