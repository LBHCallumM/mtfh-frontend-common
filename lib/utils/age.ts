import { differenceInYears, format, isValid, parseISO } from "date-fns";

export const isUnderEighteen = (dob: string) => {
  const isValidDate = parseISO(dob);
  if (!isValid(isValidDate)) {
    return true;
  }

  const dobDate = format(new Date(dob), "yyyy-MM-dd");
  const ageInYears = differenceInYears(new Date(), new Date(dobDate));
  return ageInYears < 18;
};
