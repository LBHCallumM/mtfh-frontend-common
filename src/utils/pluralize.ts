const pluralize = (word: string, value: number): string =>
  `${word}${Math.abs(value) !== 1 ? 's' : ''}`;

export default pluralize;
