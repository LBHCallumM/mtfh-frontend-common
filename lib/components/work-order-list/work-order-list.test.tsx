import React from "react";
import {
  getWorkOrdersV2,
  mockWorkOrders,
  render,
  server,
} from "@hackney/mtfh-test-utils";
import { screen, waitFor } from "@testing-library/react";
import locale from "../../locale";
import { formatDate } from "../../utils";
import { WorkOrderList } from "./work-order-list";

test("WorkOrderList renders", async () => {
  const { container } = render(<WorkOrderList assetId="00023400" />);
  const mockOrder = mockWorkOrders[0];

  await waitFor(() => {
    screen.getByText(
      `${mockOrder.tradeDescription}: ${mockOrder.description.substring(0, 50)}`,
    );
  });

  screen.getByText(formatDate(mockOrder.dateRaised));
  screen.getByText(mockOrder.priority);
  screen.getByText(mockOrder.status);

  expect(screen.getAllByText("Raised at:").length).toBe(12);
  expect(screen.getAllByText("Priority:").length).toBe(12);

  screen.getByText(locale.components.workOrderList.seeAllWorkOrders);

  expect(container).toMatchSnapshot();
});

test("WorkOrderList returns an error from the api", async () => {
  server.use(getWorkOrdersV2("error", 500));

  render(<WorkOrderList assetId="00023400" />);

  await waitFor(() => {
    screen.getByText(locale.components.workOrderList.errors.unableToFetchWorkOrder);
    screen.getByText(
      locale.components.workOrderList.errors.unableToFetchWorkOrderDescription,
    );
  });
});
