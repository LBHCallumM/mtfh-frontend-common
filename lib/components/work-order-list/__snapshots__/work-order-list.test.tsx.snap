// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkOrderList renders 1`] = `
<div>
  <div
    class="work-order-list"
  >
    <div
      class="govuk-form-group lbh-form-group"
      id="filter"
    >
      <label
        class="govuk-label lbh-label"
        for="filter-field"
      >
        Show:
        
      </label>
      <select
        class="govuk-select lbh-select"
        data-testid="work-order-list:filter"
        id="filter-field"
      >
        <option
          value="Cancelled"
        >
          repairs cancelled
        </option>
        <option
          value="Completed"
        >
          repairs completed
        </option>
        <option
          value="InProgress"
        >
          repairs in progress
        </option>
        <option
          value="Locked"
        >
          repairs locked
        </option>
        <option
          value="OnHold"
        >
          repairs on hold
        </option>
      </select>
    </div>
    <div>
      <div
        class="mtfh-card-list"
      >
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/8070582"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class="work-order-list-item__trim"
                >
                  dolorem
                  : 
                  Aliquid est quos non qui qui totam odio blanditiis
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  14/05/2001
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  aut
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              eius
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/84791138"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  nostrum
                  : 
                  Vel culpa dolore ad nihil natus aut tempore.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  27/01/2015
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  in
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              et
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/113666338"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  ea
                  : 
                  Et aut nihil doloremque deleniti et consequuntur.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  28/06/2013
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  et
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              ducimus
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/97651871"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  repellendus
                  : 
                  Nulla molestias similique.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  14/09/2016
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  consectetur
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              maiores
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/106202098"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  cumque
                  : 
                  Qui fugiat sint non.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  18/09/2006
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  nihil
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              laborum
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/40017102"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class="work-order-list-item__trim"
                >
                  voluptas
                  : 
                  Ducimus qui repellendus laudantium blanditiis aut 
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  19/10/2000
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  laudantium
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              maiores
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/76461010"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  delectus
                  : 
                  Non maxime eveniet voluptas quos et vero et quam.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  15/10/2019
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  beatae
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              itaque
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/117874276"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  dicta
                  : 
                  Autem qui et consequatur ea laborum.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  04/03/2008
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  et
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              perspiciatis
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/42349149"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  laborum
                  : 
                  Occaecati nihil sed quis.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  31/10/2010
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  omnis
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              quas
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/984924"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  facere
                  : 
                  Rem rerum ut error.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  28/11/2002
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  molestias
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              quos
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/474545"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  ipsa
                  : 
                  Reiciendis velit commodi id est quidem omnis.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  09/12/2014
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  laboriosam
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              ut
            </div>
          </div>
        </div>
        <div
          class="mtfh-link-box"
        >
          <div
            class="mtfh-card"
          >
            <div
              class="mtfh-link-overlay"
            >
              <a
                class="govuk-link lbh-link lbh-link"
                href="/api/v1/work-orders/58703263"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class=""
                >
                  aut
                  : 
                  Consequatur ut ullam quasi placeat quam.
                </span>
              </a>
            </div>
            <dl
              class="govuk-summary-list govuk-summary-list--no-border mtfh-summary-list--inline lbh-summary-list"
            >
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Raised at:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  27/07/2010
                </dd>
              </div>
              <div
                class="govuk-summary-list__row"
              >
                <dt
                  class="govuk-summary-list__key"
                >
                  Priority:
                </dt>
                <dd
                  class="govuk-summary-list__value"
                >
                  voluptatem
                </dd>
              </div>
            </dl>
            <hr
              class="mtfh-card-break"
            />
            <div
              class="work-order-list-item__status"
            >
               
              tempora
            </div>
          </div>
        </div>
      </div>
      <a
        class="govuk-link lbh-link repair-list__link"
        href="/api/v1/properties/00023400"
        rel="noopener noreferrer"
        target="_blank"
      >
        See all repairs in Repairs Hub &gt;
      </a>
    </div>
  </div>
</div>
`;
